aerich/utils.py

```diff
@@ -101,3 +103,112 @@ def import_py_file(file: Union[str, Path]) -> ModuleType:
     module = importlib.util.module_from_spec(spec)  # type:ignore[arg-type]
     spec.loader.exec_module(module)  # type:ignore[union-attr]
     return module
+
+
+def _pick_match_to_head(m2m_fields: list[dict], field_info: dict) -> list[dict]:
+    """
+    If there is a element in m2m_fields whose through or name is equal to field_info's
+    and this element is not at the first position, put it to the head then return the
+    new list, otherwise return the origin list.
+
+    Example::
+        >>> m2m_fields = [{'through': 'u1', 'name': 'u1'}, {'throught': 'u2', 'name': 'u2'}]
+        >>> _pick_match_to_head(m2m_fields, {'through': 'u2', 'name': 'u2'})
+        [{'through': 'u2', 'name': 'u2'}, {'throught': 'u1', 'name': 'u1'}]
+
+    """
+    through = field_info["through"]
+    name = field_info["name"]
+    for index, field in enumerate(m2m_fields):
+        if field["through"] == through or field["name"] == name:
+            if index != 0:
+                m2m_fields = [field, *m2m_fields[:index], *m2m_fields[index + 1 :]]
+            break
+    return m2m_fields
+
+
+def reorder_m2m_fields(
+    old_m2m_fields: list[dict], new_m2m_fields: list[dict]
+) -> tuple[list[dict], list[dict]]:
+    """
+    Reorder m2m fields to help dictdiffer.diff generate more precise changes
```

Comment:
Can you please explain **why** we need to order the fields in this way? Is the code that follows this expects some sort of ordering? What are the criteria for this ordering?

Reply:
1. why we need to order the fields in this way?
Good question~
May be a custom diff function is a better solution:
```diff
-                if old_m2m_fields and new_m2m_fields:
-                    old_m2m_fields, new_m2m_fields = reorder_m2m_fields(
-                        old_m2m_fields, new_m2m_fields
-                    )
-                for action, option, change in diff(old_m2m_fields, new_m2m_fields):
+                for action, option, change in diff_plus(old_m2m_fields, new_m2m_fields):
```
2. Is the code that follows this expects some sort of ordering? What are the criteria for this ordering?
All cases are covered by unittest: https://github.com/tortoise/aerich/pull/376/files#diff-33c13e0b177bacd2f02e29bcb8aea5b49e7ce34901fd8f41fefb65defba1bd33R7

Reply:
Thanks for the reply!

I'm not saying that a custom diff function is better, I'm just trying to understand why this ordering is required. if it's really hard to explain with a spoken language, it's a good sign that something isn't right.

Reply:
@henadzit I updated the PR description

Reply:
> @henadzit I updated the PR description

@henadzit Thanks for you comments. I created a new PR, which fix the issue by another solution: https://github.com/tortoise/aerich/pull/390